zenip-42202.md

```diff
@@ -0,0 +1,121 @@
+
+# ZenIP-42202
+
+    ZenIP: 42202
+    Title: Non-ceasable Sidechains
+    Owner: Paolo Tagliaferri, <paolotagliaferri@horizenlabs.io>; cronic, <cronic@horizenlabs.io>
+    Discussions-To: Paolo Tagliaferri, <paolotagliaferri@horizenlabs.io>
+    Comments-URI: https://horizen.global/invite/discord
+    Status: Draft
+    Type: Consensus
+    Created: 2022-08-01
+    License: MIT
+
+## Table of Contents
+
+<!--ts-->
+
+- [Abstract](#abstract)
+- [Motivation](#motivation)
+- [Specification](#specification)
+  - [Constraints](#constraints)
+  - [Maturity](#maturity)
+
+
+## Abstract
+
+This document introduces a new type of sidechain that never ceases from processing transactions. The version for this type of sidechains is “v2”.
+
+With sidechains v0 and v1, all sidechains must specify a `withdrawalEpochLength` > 0 at time of creation, this value determines the length of the epoch and when a submission window opens and closes for each epoch. 
+
+The life cycle of any sidechain is composed of epochs, that are sets of consecutive blocks whose size is determined by the `withdrawalEpochLength`. The first epoch begins with the block in which the sidechain is created (through a transaction containing its SidechainCreationOutput), the second one starts after `withdrawalEpochLength` blocks and so on.
+
+At the beginning of any epoch, except for the first one, there is a subset of blocks that constitutes the so-called **submission window**, in which the sidechain is required to publish at least one certificate for the previous epoch. Any certificate received outside this window is automatically rejected.
+
+Certificates act as an heartbeat of a sidechain and are the only way to move coins back to the mainchain.
+
+If a sidechain doesn’t provide at least one certificate for an epoch, it is considered as “ceased” and no more transactions to and from the sidechain are accepted (only *Ceased Sidechain Withdrawals* are allowed as an emergency procedure to recover funds).
+
+![A ceasable sidechain](ZenIP-42202/ZenIP-42202-1.png)
+
+We propose to introduce sidechains v2 that are not required to submit certificates during a submission window but can continue to process transactions without the risk of ceasing.
+
+Please note that this non-ceasing behavior is not enabled for all sidechains v2, but it must be explicitly requested by setting the `withdrawalEpochLength` to “0”.
+
+Sidechains v2 will have two different behaviors based on the creation parameters:
+
+- if the `withdrawalEpochLength` is set to “0”, the sidechain never ceases
+
+- if the `withdrawalEpochLength` is not “0”, the sidechain works exactly as sidechains v1 (with normal ceasing behavior)
```

Comment:
We took into account the possibility to add a new boolean in the `SidechainCreationOutput` but in the end we preferred to avoid for the following reasons:
- the new boolean would be strictly connected to the value of `withdrawalEpochLength`, meaning that if the sidechain is `ceasable` than the `withdrawalEpochLength` must be greater than zero, otherwise, for non-ceasable sidechains, only zero is accepted as valid. In this sense, adding a new field doesn't provide more flexibility but at the same time would bring more complexity as the validation of a request should take into account two fields instead of one
- adding a new boolean would add extra space to the serialization of the `SidechainCreationOutput`; even though it would be minimal, we always try to minimize the footprint of the blockchain
- when creating a sidechain, thinking about the value to be set for the `withdrawalEpochLength` is always crucial, also for ceasable sidechains, as it determines the frequency of certificates to be published. Moreover, when creating a non-ceasable sidechain, the validation rules always enforce that CSW and MBTR are disabled, so it seems unlucky to create a sidechain with wrong behavior by mistake.
- with a new boolean, we should also change the interface of the RPC command to create sidechains, making unusable any existing script or tools used by other teams unless they get properly updated

Reply:
I see, thanks for the explanation! 
What about making `withdrawalEpochLength` optional? so that it is only added to the `SidechainCreationOutput` if it is a ceasable sidechain?

Reply:
You're welcome!

Currently the RPC command `sc_create` considers `withdrawalEpochLength` as an optional field, whose default value is 100 (that is the minimum epoch length for ceasable sidechains).
Changing such behavior could break the compatibility with tools and scripts that are eventually calling `sc_create` without explicitly setting the `withdrawalEpochLength`.
Making the change only to the `SidechainCreationOutput` could cause similar issues, as tools like the Explorer and the Bootstrapping Tool might be implemented to always parse the `withdrawalEpochLength` from the creation transaction. In general, we try to reduce as much as possible changes on the mainchain that would affect other software.
There is also a serialization issue here: if we make that field optional, we wouldn't have a way to understand whether the `withdrawalEpochLength` has to be read or not. To make things even more complex, in the past, when we introduced sidechains v1, we had to use the first byte of the `withdrawalEpochLength` to specify the sidechain version, so making changes to the serialization of this field would have impacts also on the `version` field.

Reply:
I see! Ok sounds like allowing 0 value is the best option then 😄